src/nyx/features/gabor.cpp

```diff
@@ -552,7 +913,18 @@ void GaborFeature::reduce (size_t start, size_t end, std::vector<int>* ptrLabels
         LR& r = (*ptrLabelData)[lab];
 
         GaborFeature gf;
-        gf.calculate (r);
+
+
+        #ifdef USE_GPU

```

Comment:
This method of selective calculating a feature on CPU- or GPGPU-side implemented in Gaborfeature::reduce() that I originally suggested you has become old. There is a better one. The problem with the old method is that if Gaborfeature::reduce() is called from multiple CPU-side threads (command line argument --reduceThreads=n, n>1), function GaborFeature::GaborEnergyGPU() will be hitting the GPU resource in parallel creating shortage of thread blocks. The better method to call Gabor feature calculation mentioned is to put the #ifdef USE_GPU and theEnvironment.using_gpu() logic in the reduction function and call the GPU code in the main thread keeping the CPU code callable in multiple threads. You can check file https://github.com/PolusAI/nyxus/blob/main/src/nyx/reduce_trivial_rois.cpp, lines 220-241, for an example of implementing this logic for 2D moments. Feel free to message me for more details.